pybandits/simulator.py

```diff
@@ -0,0 +1,592 @@
+    pydantic_version,
+)
+from pybandits.utils import visualize_via_bokeh
+
+
+class Simulator(PyBanditsBaseModel, ABC):
+    """
+    Simulate environment for multi-armed bandits.
+
+    This class performs simulation of Multi-Armed Bandits (MAB). Data are processed in batches of size n>=1.
+    Per each batch of simulated samples, the mab selects one a1nd collects the corresponding simulated reward for
+    each sample. Then, prior parameters are updated based on returned rewards from recommended actions.
+
+    Parameters
+    ----------
+    mab : BaseMab
+        MAB model.
+    n_updates : PositiveInt, defaults  to 10
+        The number of updates (i.e. batches of samples) in the simulation.
+    batch_size: PositiveInt, defaults to 100
+        The number of samples per batch.
+    probs_reward : Optional[pd.DatafFame], default=None
+        The reward probability for the different actions. If None probabilities are set to 0.5.
+        The keys of the dict must match the mab actions_ids, and the values are float in the interval [0, 1].
+        e.g. probs_reward={'a1': 0.6, 'a2': 0.8, 'a3': 1.}
+    save : bool, defaults to False
+        Boolean flag to save the results.
+    path : string, default to ''
+        Path where_results are saved if save=True
+    file_prefix : string, default to ''
+        Prefix for the file name where results are saved.
+    random_seed : int, default=None
+        Seed for random state. If specified, the model outputs deterministic results.
+    verbose :  bool, default=False
+        Enable verbose output. If True, detailed logging information about the simulation are provided.
+    visualize : bool, default=False
+        Enable visualization of the simulation results.
+    """
+
+    mab: BaseMab
+    n_updates: PositiveInt = 10
+    batch_size: PositiveInt = 100
+    probs_reward: Optional[pd.DataFrame] = None
+    save: bool = False
+    path: str = ""
+    file_prefix: str = ""
+    random_seed: NonNegativeInt = None
+    verbose: bool = False
+    visualize: bool = False
+    _results: pd.DataFrame = PrivateAttr()
+    _base_columns: List[str] = PrivateAttr()
+    _cumulative_col_prefix: str = "cum"
+    # Define dash patterns, markers, and colors for lines
+    _dash_patterns = ["solid", "dashed", "dotted"]
+    _markers = ["circle", "square", "triangle", "diamond", "star"]
+
+    if pydantic_version == PYDANTIC_VERSION_1:
+
+        class Config:
+            arbitrary_types_allowed = True
+            allow_population_by_field_name = True
+
+    elif pydantic_version == PYDANTIC_VERSION_2:
+        model_config = {"arbitrary_types_allowed": True, "populate_by_name": True}
+    else:
+        raise ValueError(f"Unsupported pydantic version: {pydantic_version}")
+
+    @field_validator("probs_reward", mode="before")
+    @classmethod
+    def validate_probs_reward_values(cls, value):
+        if value is not None:
+            if not value.applymap(lambda x: 0 <= x <= 1).all().all():
```

Comment:
Small suggestion.
Instead of implementing this login manually (all values in [0,1]), you can use pydantic with confloat or even better Annotated

https://docs.pydantic.dev/latest/api/types/#pydantic.types.confloat

Reply:
This is a good suggestion, but to the best of my understanding, because probs_reward is a dataframe, adding constraints on its values is not supported by pydantic. Am I missing something?

Reply:
Indeed that's right!